### PR TITLE
fix schema to allow for custom blocks in theme.json

### LINKED ISSUE #36338: JSON Schema of `theme.json` doesn't support custom blocks
### Description

When using the JSON Schema from the `theme.json` file located at https://schemas.wp.org/trunk/theme.json the `blocks` key only supports all the different core blocks right now. https://github.com/WordPress/gutenberg/blob/a9b7e98ecf48ae134e42074abb23115fd31c9c7a/schemas/json/theme.json#L566

I think the best option here is to add the blocks in a `anyOf` block where it can either be one of the defined core blocks or it can be a `patternProperty` that needs to follow the `namespace/block-name` pattern and then also gets the schema applied to the properties within. 

### Step-by-step reproduction instructions

Add the "$schema" key to a `theme.json` like so:

```json
{
    "$schema": "https://schemas.wp.org/trunk/theme.json"
}
```

In the `settings.blocks` object try to add a custom block like so:

```json
{
    "$schema": "https://schemas.wp.org/trunk/theme.json",
    "settings": {
        "blocks": {
            "example/block": {
                ...
            }
         }
     }
}
```

You will see that it gets flagged as incorrect. 

### Screenshots, screen recording, code snippet

_No response_

### Environment info

_No response_

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by fabiankaegy**
## Description

This PR fixes the JSON Schema to add support for Custom blocks in the `blocks` section of `theme.json` files. It used the regular expression from here: https://github.com/WordPress/gutenberg/blob/984160d1368695f22222aea4e20527d7699b842c/packages/blocks/src/api/registration.js#L280 to check for the validity of the block name.

Closing #36338

## How has this been tested?

```json
{
	"$schema": "../theme.json",
	"version": 1,
	"settings": {
		"blocks": {
			"example/namespace": {
				"color": {
					"background": true
				}
			},
			"core/code": {},
		}
	}
}
```

## Types of changes

Update JSON Schema to allow custom blocks

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal).

### CODE REVIEW

File: schemas/json/theme.json [hunk 1]
```diff
@@ -312,258 +312,270 @@
...
+						"core/widget-group": {
+							"$ref": "#/definitions/settingsPropertiesComplete"
+						}
+					}
 				},
-				"core/widget-group": {
-					"$ref": "#/definitions/settingsPropertiesComplete"
+				{
+					"patternProperties": {
+						"^[a-z][a-z0-9-]*\/[a-z][a-z0-9-]*$": {
```
Comment (gziolo) [inline schemas/json/theme.json:571]: I wasn't aware that it has to start with a letter 😄 

I guess the best part of working on those schemas is that we are able to clarify what edge cases should be.

**Review by gziolo — COMMENTED** [review]
Maybe we should avoid hardcoding the list of all core blocks and use only the pattern? If there would be ever anything special about core blocks, we can always channel it through another regular expression that uses `core` as a namespace.

By the way, we can also be more strict about the block name in `block.jscon` schema:

https://github.com/WordPress/gutenberg/blob/db67d94b08135ed8c4282cc2c69726225f704dbe/schemas/json/block.json#L24-L27

**Comment by fabiankaegy** [comment]
> By the way, we can also be more strict about the block name in `block.jscon` schema:
> 
> https://github.com/WordPress/gutenberg/blob/db67d94b08135ed8c4282cc2c69726225f704dbe/schemas/json/block.json#L24-L27

I have added a separate PR for that here: #36343

**Comment by fabiankaegy** [comment]
> Maybe we should avoid hardcoding the list of all core blocks and use only the pattern? If there would be ever anything special about core blocks, we can always channel it through another regular expression that uses core as a namespace.


@gziolo I'm not sure why they were all added specifically. But it comes with one nice side-effect and that is automatic code completion for all of the core blocks. When you start typing in the `blocks` section you directly get the suggestion of all the core blocks. Which is kind of nice. 

But it feels like another place where we need to manually keep track of the list of blocks. Which can get missed very easily and then it gets frustrating.

File: schemas/json/theme.json (hunk 1 above)
Comment (fabiankaegy) [reply schemas/json/theme.json:571]: Yeah agreed. Working on these schemas really helps surface some things that I was previously just not aware of like this :)

**Comment by gziolo** [comment]
> But it feels like another place where we need to manually keep track of the list of blocks. Which can get missed very easily and then it gets frustrating.

Automatic code completion sounds like something worth an effort. We can always write integration tests that enforce that new block names are included in the `theme.json` schema. We do it already for block test fixtures. Let's hear back from @mkaz and @ajlende about the purpose of the list of block names.

**Comment by ajlende** [comment]
Yeah, I had them all added for automatic code completion. I figured that the core blocks wouldn't be changing that frequently, so the maintenance wouldn't be too much of a hassle. Adding tests to make sure that it stays up-to-date sounds like a good plan!

**Review by gziolo — APPROVED** [review]

**Comment by fabiankaegy** [comment]
@gziolo What can we do about the failed checks? It is preventing the merge

**Comment by gziolo** [comment]
Those failures are unrelated. I ignored them 😃

**Pull request merged**